### PR TITLE
1.5/1227 create cli mvl report runner 2

### PR DESCRIPTION
**Pull request by charlesstrange2**

### CODE REVIEW

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        'Payment Request Date',
+        'Trader Name',
+        'Retail Outlet',
+        'Reimbursed Date'
+    ];
+
+    /**
+     * The report's query template
+     *
+     * TODO: refactor this as eloquent lookups when it's finalised.
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: Will you do the refactor in this PR - or mean at a later date? You know I'm not keen on raw SQL. :smile:

Comment (charlesstrange2) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: At a later date. The sheer complexity of this means that SQL is probably the most efficient way to pull this particular report. I'd like it to be refactored, as the functions might be useful in a future "admin" sprint.

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+SELECT
+  sponsors.name AS 'Area',
+  vouchers.code AS 'Voucher Code',
+  dispatch_date AS 'Dispatch Date',
+  disbursed_at AS 'Disbursed Date',
+  rvid AS 'RVID',
+  pri_carer_name AS 'Participant',
+  payment_request_date AS 'Payment Request Date',
+  trader_name AS 'Trader Name',
+  market_name AS 'Retail Outlet',
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: lol 'Retail Outlet'

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 1 above)
Comment (katjam) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: Yeah - it looks nuts. Agree that functions will likely turn out to be useful but I guess you are in a rush?

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    LEFT JOIN markets ON traders.market_id = markets.id
+  ) AS markets_query
+    ON markets_query.id = vouchers.trader_id
+
+  # Pivot dispatched date
+  LEFT JOIN (
+    SELECT voucher_states.voucher_id,
+           voucher_states.created_at AS dispatch_date
+    FROM voucher_states
+    WHERE voucher_states.`to` = 'dispatched'
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: Do you need the backticks?

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: time - money in this case.

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 3 above)
Comment (charlesstrange2) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: yes, 'to' is a reserved word in mysql and, unfortunatly the column is called that (grr, not my choice). the backticks required to be certain no strangeness occurs

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+          SELECT MIN(id) as id
+          FROM carers t3
+          GROUP BY t3.family_id
+        ) t2 ON t2.id = t1.id 
+      ) AS pri_carer_query
+        ON families.id = pri_carer_query.family_id
+  ) AS rvid_query
+    ON rvid_query.id = vouchers.bundle_id
+;
+EOD;
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: I've checked for obvious typs - but pretty hard to follow in terms of making sure no relevant data is being lost - by just reading the queries. Have you stepped through it to make sure? I'd be easier to read - but probably slower to iterate through collections and relations building up the data you need.

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+      LEFT JOIN centres ON families.initial_centre_id = centres.id
+        
+      # Need to join the Primary Carer here; Primary Carers are only relevant via bundles.
+      LEFT JOIN (
+        
+        # We need the *first*, by self join grouping (classic technique, so SQLite can cope)
+        SELECT t1.name, t1.family_id
+        FROM carers t1
+        INNER JOIN (
+          SELECT MIN(id) as id
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: Will soft deletes or reinstated main carers be an issue here? I can't remember how you settled on designating Main Carer.

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+
+namespace App\Console\Commands;
+
+use Carbon\Carbon;
+use DB;
+use Excel;
+use Illuminate\Support\Facades\Storage;
+use Illuminate\Console\Command;
+use Log;
+use Maatwebsite\Excel\Writers\LaravelExcelWriter;
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: Nice - is this the replacement for deprecated phpexcel package we used before?

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 3 above)
Comment (katjam) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: Nice.

Comment (katjam) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: I'm not a big fan of that state machine. :1st_place_medal: Good wrangling though!

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 6 above)
Comment (katjam) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: We really should upgrade. Even 5.4 been out of security fixes for a year already. https://laravel-news.com/laravel-release-process

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 5 above)
Comment (katjam) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: Sounds good to me.

File: app/Console/Commands/CreateMasterVoucherLogReport.php (hunk 2 above)
Comment (charlesstrange2) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: Personally i'd consider the trader stall that, rather than the market as a whole, but we'll use ARC's terms.

File: app/Console/Commands/CreateMasterVoucherLogReport.php [hunk 7]
```diff
@@ -73,16 +75,19 @@ class CreateMasterVoucherLogReport extends Command
      */
     private $report = <<<EOD
 SELECT
+  vouchers.code AS 'Voucher Number',
+  # To be determined, we don't know this yet.
+  '' AS 'Distributed to',
```
Comment (katjam) [inline app/Console/Commands/CreateMasterVoucherLogReport.php]: Neat. Does that work?

Comment (charlesstrange2) [reply app/Console/Commands/CreateMasterVoucherLogReport.php]: yep, you can put constant strings in columns, so en empty string is treated the same.
